### PR TITLE
Implement RNNT loss

### PR DESCRIPTION
**Pull request by cong-or**
Adds RNN Transducer (RNNT) loss to complete the second half of #4459 (CTC was merged in #4529).                                                                                             
                                                                
The implementation computes the forward variable over the (T, U) alignment lattice in log-space, following the algorithm from [Graves 2012](https://arxiv.org/abs/1211.3711). Backward gradients come from Burn's autodiff — no manual gradient code needed.            
                                                                                                                                                                                              
  ### Changes                                                                      

  - `crates/burn-nn/src/loss/rnnt.rs` — `RNNTLossConfig` / `RNNTLoss` with `forward` and `forward_with_reduction`
  - `crates/burn-nn/src/loss/mod.rs` — module registration
  - `burn-book/src/building-blocks/module.md` — added to the Loss table

  ### API

  ```rust
  let rnnt = RNNTLossConfig::new().init(); // blank=0, fused_log_softmax=true

  // logits: [B, T, U+1, V] from joiner network
  let loss = rnnt.forward(logits, targets, logit_lengths, target_lengths);
  ```
  fused_log_softmax controls whether log_softmax is applied internally (like CTC's behavior) or expects pre-computed log-probs.

  Tests

  - Unit tests with hand-derived expected values (uniform probs, empty target)
  - Comparison tests against torchaudio.functional.rnnt_loss checking both forward loss and backward gradients
  - Variable-length masking verified by asserting zero gradients in padded regions

  Closes the RNNT portion of #4459.

### CODE REVIEW

File: crates/burn-nn/src/loss/rnnt.rs [hunk 1]
```diff
@@ -0,0 +1,836 @@
...
+/// Configuration for [RNNTLoss](RNNTLoss).
+#[derive(Config, Debug)]
+pub struct RNNTLossConfig {
+    /// Index of the blank label in the vocabulary. Default: `0`.
+    #[config(default = 0)]
+    pub blank: usize,
+    /// If `true`, applies `log_softmax` to the last dimension internally.
+    /// If `false`, the input must already be log-probabilities. Default: `true`.
+    #[config(default = true)]
+    pub fused_log_softmax: bool,
```
Comment (laggui) [inline crates/burn-nn/src/loss/rnnt.rs]: I see this matches the args to torchaudio's RNNT loss, but to keep it consistent with other losses / metrics I would rename it to `logits`.

Essentially, treat the inputs as logits, applying log-softmax when computing the loss (equivalent to `fused_log_softmax = true`. If `false`, inputs are assumed to already be log-probabilities.

File: crates/burn-nn/src/loss/rnnt.rs [hunk 2]
```diff
@@ -0,0 +1,836 @@
...
+        for t in 1..max_t {
+            let new = self.step_alpha(&alpha, &lpb, &lpl, t);
+            let new = neg_inf.clone().mask_where(u_mask.clone(), new);
+
+            // Keep old alpha for samples where t >= logit_lengths[b]
+            let valid = logit_lengths
+                .clone()
+                .greater_elem(t as i64)
+                .reshape([b, 1])
+                .expand([b, max_up1]);
```
Comment (laggui) [inline crates/burn-nn/src/loss/rnnt.rs]: 1. If it's supposed to be for `>= logit_lengths`, you should use `greater_equal_elem`

2. The only thing that changes is `t` for the comparison, so you could have reshape + expand outside the loop and just apply `greater_equal_elem` for the mask in the loop

File: crates/burn-nn/src/loss/rnnt.rs [hunk 3]
```diff
@@ -0,0 +1,836 @@
...
+        assert!(
+            (a - e).abs() < tol,
+            "[{}] expected {:.6}, got {:.6} (diff {:.6})",
+            i,
+            e,
+            a,
+            (a - e).abs()
+        );
+    }
+}
```
Comment (laggui) [inline crates/burn-nn/src/loss/rnnt.rs]: Should use `tensor.into_data().assert_approx_eq(...)` instead

File: crates/burn-nn/src/loss/rnnt.rs (hunk 1 above)
Comment (cong-or) [reply crates/burn-nn/src/loss/rnnt.rs]: Done, renamed to logits with updated docs to match the pattern in BinaryCrossEntropyLoss and CrossEntropyLoss.

File: crates/burn-nn/src/loss/rnnt.rs (hunk 2 above)
Comment (cong-or) [reply crates/burn-nn/src/loss/rnnt.rs]: On `greater_equal_elem` — the code was correct, the comment was just misleading. valid is where we update alpha, so we need `logit_lengths > t`. With >=, shorter sequences process one extra invalid step — `variable_lengths_b3` fails with loss values off by ~1.5. 

I Rewrote the comment to be clearer.

**Comment by cong-or** [comment]
> Thanks for contributing this implementation!
> 
> I only have a couple of minor comments

@laggui Thanks for the review. I have addressed the issues raised.

File: crates/burn-nn/src/loss/rnnt.rs (hunk 3 above)
Comment (cong-or) [reply crates/burn-nn/src/loss/rnnt.rs]: Done, switched to assert_approx_eq and dropped the custom assert_close helper.

**Comment by laggui** [comment]
> @laggui Thanks for the review. I have addressed the issues raised.

Thanks for addressing my comments!

Looks like the code needs to run through `cargo fmt` first. Should be good afterwards.

**Comment by cong-or** [comment]
> > @laggui Thanks for the review. I have addressed the issues raised.
> 
> Thanks for addressing my comments!
> 
> Looks like the code needs to run through `cargo fmt` first. Should be good afterwards.

thanks - just ran cargo fmt, hopefully ci will be green!

**Comment by cong-or** [comment]
> LGTM!
> 
> Failing macos test is not due to your changes, should be related to the latest cubecl changes.

Thanks for the review! Let me know if there's anything else needed on my end.
